### PR TITLE
Feature/ubuntu

### PR DESCRIPTION
**Pull request by mmiesch**
This is a major re-organization of the repo, preparing it for use on multiple platforms.  Currently only the minimal jedi stack is tested on AWS and it is sufficient to run ufo-bundle.  Future development will include the optional packages and will adapt it for use with the containers and HPC systems.  

Please read the README file in the top level of the repo to get an idea of how I intend to use this

### CODE REVIEW

**Comment by aerorahul** [comment]
quick glance comment:
1. eckit is installing ecbuild along with it.  It would be better to have a separate build for ecbuild and its associated modulefile.

**Comment by mmiesch** [comment]
Thanks @aerorahul for your comments - I just did a push to separate ecbuild and eckit (good idea!) and I updated and checked the nccmp.sh script - it looks like it's working.  I also went through the entire thing again from scratch and found a few bugs.  I still have not updated the scripts for the non-essential libs (e.g. fftw, esmf...) but that can be done in a future PR.  And yes, at some point I plan to restore MacOS functionality.  That's harder to test, though, because I don't really want to mess up my current Mac configuration.  I left all the DYLD_LIBRARY_PATHs in so it should be straightforward.  It may still work out of the box if you just do the stuff in setup_environment.sh manually.

**Comment by mmiesch** [comment]
@aerorahul - Sorry, which comment below?

**Comment by aerorahul** [comment]
@mmiesch 
Sorry about that.

I notice in `compiler` modulefiles, the default for base is `/opt` as opposed to `/opt/modules`
E.g. see in the `netcdf` modulefile. Same for hdf5, esmf.  These can be built without MPI.
I understand that the builds are building compiler+mpi, and not compiler only.

Your `nccmp` modification requires loading MPI, yet, the package is placed in $compilername-$compilerversion and not under $compilername-$compilerversion/$mpiname-$mpiversion.

**Comment by mmiesch** [comment]
Thanks @aerorahul - Yes, I was debating about where to put nccmp.  If I leave out that load of mpi then it does not have the proper module path setup to find hdf5 and thus netcdf.  So, I think I have to include it.  Then I thought I should put nccmp under the mpi module path but I second-guessed myself again, saying that a comparison of netcdf files should not depend on the mpi implementation.  So, I'm going back and forth - do you think I should move it to mpi?  And, yes, I know hdf5 and netcdf can in principle be built without mpi support but one of the motivations for this repo is to use common build options across platforms and I don't see many jedi use cases where we would *not* build in the mpi support.  I'll fix the /opt thing.

**Comment by aerorahul** [comment]
@mmiesch 
To my knowledge none of the JEDI core applications use the parallel functionality of HDF5 and netCDF.

Having said that, if `nccmp` is built with mpi, its module and the package should be under mpi.  This will keep the `lmod` module hierarchy intact.  `build_nccmp.sh` should be updated and be allowed to be built with or without mpi. See `build_hdf5.sh` for the necessary changes. FC, CC variables will have to be updated to use the mpi compiler when mpi is loaded. `prefix` will contain mpi variables.  Same will be applicable to `build_nco.sh`, because you won't have a `serial` version of `netcdf`, and when `cdo` is built, same thing.  

In my opinion, since `netcdf` and `hdf5` are so trivial and fast to build, why not keep a `serial` version handy?  They don't take much space or and almost no effort to maintain.

**Comment by mmiesch** [comment]
Thanks @aerorahul - I agree that jedi apps don't *currently* make use of the parallel netcdf functionality - for now - but we wanted to build that in to have it available in the future because we expect that we will be using it - possibly soon.  For now I'll just put in in the mpi tree and maybe in the future we can add a serial version.
One question - in your build_nccmp script you look to see if "netcdf_c" is there to do a header pad comparison.  However, if it is there, it would be called "netcdf_c-$version", wouldn't it?  I'm not sure how to gracefully handle that - maybe by looking for the netcdf.h file in the pkg directory and then getting the full directory name from that?

File: buildscripts/libs/build_nccmp.sh [hunk 1]
```diff
@@ -37,7 +37,7 @@ cd ${JEDI_STACK_ROOT}/${PKGDIR:-"pkg"}
 [[ -d build ]] && rm -rf build
 mkdir -p build && cd build
 
-prefix="${PREFIX:-"/opt/modules"}/$compiler/$name/$version"
+prefix="${PREFIX:-"/opt/modules"}/$compiler/$mpi/$name/$version"
```
Comment (aerorahul) [inline buildscripts/libs/build_nccmp.sh]: $mpi is not defined anywhere.
It needs to be obtained from $MPI similar to L12 for compiler.

**Comment by aerorahul** [comment]
once netcdf module is loaded, you will have access to environment variable $NETCDF_VERSION. The `netcdf-c` should be replaced with `netcdf-c-$NETCDF_VERSION`.

**Comment by mmiesch** [comment]
Sure - done

**Comment by mmiesch** [comment]
Thanks @aerorahul - good idea about $NETCDF_VERSION

**Comment by aerorahul** [comment]
@mmiesch 
all looks good.
merge away when ready!

File: buildscripts/libs/build_pnetcdf.sh [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+# Hyphenated version used for install prefix
+compiler=$(echo $COMPILER | sed 's/\//-/g')
+mpi=$(echo $MPI | sed 's/\//-/g')
+
+[[ $USE_SUDO =~ [yYtT] ]] && export SUDO="sudo" || unset SUDO
+
+set +x
+source $MODULESHOME/init/bash
+module load jedi-$COMPILER
+module load jedi-$MPI
```
Comment (aerorahul) [inline buildscripts/libs/build_pnetcdf.sh]: the modulefile for pnetcdf indicates it depends on hdf5.
You probably need hdf5 module loaded to be able to build pnetcdf.
I have not tested pnetcdf, but that is a hunch.

File: modulefiles/mpi/compilerName/compilerVersion/mpiName/mpiVersion/pnetcdf/pnetcdf.lua.tmpl [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+local hierA        = hierarchyA(pkgNameVer,2)
+local mpiNameVer   = hierA[1]
+local compNameVer  = hierA[2]
+local mpiNameVerD  = mpiNameVer:gsub("/","-")
+local compNameVerD = compNameVer:gsub("/","-")
+
+conflict(pkgName)
+
+always_load("hdf5")
+prereq("hdf5")
```
Comment (aerorahul) [inline modulefiles/mpi/compilerName/compilerVersion/mpiName/mpiVersion/pnetcdf/pnetcdf.lua.tmpl]: you will need it during build too.

File: modulefiles/compiler/compilerName/compilerVersion/lapack/lapack.lua.tmpl [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+setenv("LAPACK_PATH", base)
+setenv("LAPACK_DIR", base)
+setenv("LAPACK_INCLUDES", pathJoin(base,"include"))
+setenv("LAPACK_LIBRARIES", pathJoin(base,"lib"))
+setenv("LAPACK_VERSION", pkgVersion)
+
+whatis("Name: ".. pkgName)
+whatis("Version: " .. pkgVersion)
+whatis("Category: library")
+whatis("Description: SZip library")
```
Comment (aerorahul) [inline modulefiles/compiler/compilerName/compilerVersion/lapack/lapack.lua.tmpl]: Should be `Linear Algebra PACKage`, not `SZip`

File: buildscripts/libs/build_netcdf.sh [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+[[ $USE_SUDO =~ [yYtT] ]] && export SUDO="sudo" || unset SUDO
+
+set +x
+source $MODULESHOME/init/bash
+module load jedi-$COMPILER
+module load jedi-$MPI
+module load szip
+module load hdf5
+module load pnetcdf
```
Comment (aerorahul) [inline buildscripts/libs/build_netcdf.sh]: There is a circular dependency here.
pnetcdf4 compilation has a flag -with-netcdf4, which requires netcdf4
and here, netcdf4 is being built with pnetcdf.

pnetcdf and netcdf4 should be treated as two different independent packages. May be pnetcdf can depend on netcdf4, but the other way around is usually not done in most applications.

both implement parallel capability, and the differences are in how they use MPI-IO.
AFAIK, most modeling systems use either/or, not both.

Something to keep in mind.  Debugging can get real messy when parallel IO is involved.

**Comment by aerorahul** [comment]
@mmiesch 
Sorry for piling on more comments.  I hope these are the last.

**Comment by mmiesch** [comment]
@aerorahul - no need to apologize!  I appreciate the comments, particularly with regard to pnetcdf.  As I understand it, using the -enable-pnetcdf configure option for NetCDF-C tells NetCDF to [use pnetcdf for "classic files (CDF-1, 2, and 5)"](https://www.unidata.ucar.edu/software/netcdf/docs/getting_and_building_netcdf.html#configure_options).  The implication is that HDF5 is still used for NetCDF-4 files.  Furthermore, in order to use this option, PNETCDF needs to be installed before NetCDF.  So, that's why I omitted the --with-netcdf4 option on PNETCDF and built it first.  Furthermore, as I understand it, PNETCDF does not use HDF5 so there's no need to load hdf5 before installing PNETCDF.  And, finally, according to the [PNETCDF page](https://trac.mcs.anl.gov/projects/parallel-netcdf), "Applications can use PnetCDF library completely without NetCDF library".  So, NetCDF does not need to be installed before PNETCDF.  So, in short, I think the way it is implemented here is ok.

**Comment by mmiesch** [comment]
@aerorahul - I agree that hdf5 should be removed from the pnetcdf module file.  I just fixed that and the LAPACK comment - Thanks again!

**Comment by mmiesch** [comment]
@aerorahul - is it ok if I merge this?  We can tweak the netcdf/pnetcdf configuration later if and when we run into problems.

**Comment by aerorahul** [comment]
@mmiesch Yes. Please.

**Pull request merged**